File: NOTES/LinkedList.java
public class LinkedList {
    Node head;

    public static class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
            next = null;
        }
    }

    public boolean isEmpty() {
        return head == null;
    }

    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    public void insertAtEnd(int data) {
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode; 
            return;
        }

        newNode.next = null;

        Node last = head;

        while (last.next != null) {
            last = last.next;
        }

        last.next = newNode;
    }

    public void insertAt(int index, int data) {
        if (index == 0) {
            insertAtBeginning(data);
        }

        if (index > 0) {
            Node newNode = new Node(data);
            Node currNode = head;

            while (index > 1) {
                currNode = currNode.next;
                index--;
            }

            Node prevNode = currNode;
            Node nextNode = currNode.next;

            prevNode.next = newNode;
            newNode.next = nextNode;

        }
    }

    public void insertAfter(Node prevNode, int data) {
        if (prevNode == null) {
            return;
        }

        Node newNode = new Node(data);
        newNode.next = prevNode.next;
        prevNode.next = newNode;
    }

    public void removeAtIndex(int index) {
        // https://medium.com/swlh/implementing-linked-lists-with-java-25a4a708b5fc
        if (head == null) {
            return;
        }

        Node curr = head;

        if (index ==0) {
            head = curr.next;
            return;
        }

        for (int i = 0; curr != null && i < index-1; i++) {
            curr = curr.next;
        }

        if (curr == null || curr.next == null) {
            return;
        }

        Node nextNode = curr.next.next;

        curr.next = nextNode;
    }

    public Node search(int key) {
        Node curr = head;
        while (curr != null) {
            if (curr.data == key) {
                return curr;
            }
            curr = curr.next;
        }
        return null;
    }

    public void print() {
        Node curr = head;
        
        System.out.print("List: ");
        while (curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        Node one = new Node(10);
        Node two = new Node(20);
        Node three = new Node(30);

        list.head = one;
        one.next = two;
        two.next = three;
        
        list.print(); // 10 20 30
        
        list.insertAtBeginning(0);

        list.print(); // 0 10 20 30 
        
        list.insertAtEnd(40);
            
        list.print(); // 0 10 20 30 40

        list.insertAtEnd(50);

        list.print(); // 0 10 20 30 40 50

        list.insertAt(1, 15);

        list.print(); // 0 15 10 20 30 40 50

        list.insertAfter(two, 25);

        list.print(); // 0 15 10 20 25 30 40 50

        System.out.println(list.search(30)); // returns particular Node

        list.removeAtIndex(3);

        list.print(); // 0 15 10 25 30 40 50
    }
}

File: NOTES/linked_list.py
class Node:
    data = None
    next_node = None

    # constructor
    def __init__(self, data):
        self.data = data
    
    def __repr__(self):
        return "<Node data: %s>" % self.data

class LinkedList:
    """
    Singly Linked List
    """
    def __init__(self):
        self.head = None

    def __repr__(self):
        """
        return string representation of list in O(n) time
        """

        nodes = []
        current = self.head

        while current:
            if current is self.head:
                nodes.append("[Head: %s]" % current.data)
            elif current.next_node is None:
                nodes.append("[Tail: %s]" % current.data)
            else:
                nodes.append("[%s]" % current.data)

            current = current.next_node
        
        # display as connection of nodes
        return '-> '.join(nodes) 

    def is_empty(self):
        return self.head == None

    def size(self):
        """
        return number of nodes in list
        """
        current = self.head
        count = 0

        while current:
            count += 1
            current = current.next_node

            return count
    
    def add(self, data):
        """
        add new node to head of list in O(1) time
        """
        new_node = Node(data)
        new_node.next_node = self.head
        self.head = new_node

    def insert(self, data, index):
        """
        inserts new node at given index
        """
        if index == 0:
            self.add(data)
        
        if index > 0:
            new_node = Node(data)

            position = index
            current = self.head

            while position > 1:
                current = new_node.next_node
                position -= 1
            
            prev_node = current
            next_node = current.next_node

            prev_node.next_node = new_node
            new_node.next_node = next_node

    def remove(self, key):
        current = self.head
        prev_node = None
        found = False

        while current and not found:
            if current.data == key and current is self.head:
                found = True
                self.head = current.next_node
            elif current.data == key:
                found = True
                prev_node.next_node = current.next_node
            else:
                prev_node = current
                current = current.next_node

        return current

    def node_at_index(self, index):
        """
        Returns the Node at specified index
        Takes O(n) time
        """

        if index >= self.__count:
            raise IndexError('index out of range')

        if index == 0:
            return self.head

        current = self.head
        position = 0

        while position < index:
            current = current.next_node
            position += 1

        return current       

    def search(self, key):
        """
        return first node with data matching given key in O(n)
        otherwise returns 'None' if not found
        """
        current = self.head

        while current:
            if current.data == key:
                return current
            else:
                current = current.next_node

        return None